builtin/common/class.lua

```diff
@@ -0,0 +1,119 @@
+class = {}
+
+function class.new(name, ...)
+	local cls = {}
+
+	-- Copy inherited classes in reverse order to make the first ones overwrite later ones
+	local bases = {...}
+	for i = #bases, 1, -1 do
+		for k, v in pairs(bases[i]) do
+			cls[class.copy(k)] = class.copy(v)
+		end
+	end
+
+	cls.__name = name
+	cls.__index = cls
+
+	-- Add instance information
+	cls.__is = {[cls] = true}
+	for _, base in ipairs(bases) do
+		for base in pairs(base.__is) do
+			cls.__is[base] = true
+		end
+	end
+
+	-- Create pretty class constructor
+	setmetatable(cls, {
+		__call = function(cls, ...)
+			local args = {...}
+			if #args == 1 and class.is(args[1], cls) then
+				return class.copy(args[1])
+			else
+				return class.init(cls, ...)
+			end
+		end
+	})
+
+	return cls
+end
+
+function class.instantiate(cls)
+	return setmetatable({}, cls)
+end
+
+function class.init(cls, ...)
+	local self = class.instantiate(cls)
```

Comment:
I don't like this at all, as it requires unnecessary tablecopies: I like to have constructors like `MyClass{a = 1}`, which directly set the metatable for the passed table.